openl3/core.py

```diff
@@ -102,10 +103,77 @@ def _preprocess_audio_batch(audio, sr, center=True, hop_size=0.1):
+    S -= S.max()
+    return S
+
+
+def _librosa_mel_frontend(audio, sr=48000, n_mels=128, n_fft=2048, hop_length=242,
+                          db_amin=1e-10, db_ref=1.0, dynamic_range=80.0):
+    '''Librosa mel frontend designed to match original Kapre (0.1.4).'''
+    S = librosa.feature.melspectrogram(
+        audio, sr, n_mels=n_mels, n_fft=n_fft, hop_length=hop_length, 
+        center=True, power=1.0)
+    S = librosa.power_to_db(S, ref=db_ref, amin=db_amin, top_db=dynamic_range)
+    S -= S.max()
+    return S
+
+
+def preprocess_audio(audio, sr=48000, hop_size=0.1, input_repr=None, center=True, **kw):
+    """
+    Preprocess the audio into a format compatible with the model.
+
+    Parameters
+    ----------
+    audio : np.ndarray [shape=(N,) or (N,C)] or list[np.ndarray]
+        1D numpy array of audio data or list of audio arrays for multiple
+        inputs.
+    sr : int or list[int]
+        Sampling rate, or list of sampling rates. If not 48kHz audio will
+        be resampled.
+    hop_size : float
+        Hop size in seconds.
+    input_repr : str or None
+        Spectrogram representation used for model.
+        If input_repr, is None, then no spectrogram is computed and
+        it is assumed that the model contains the details about 
+        the input representation.
+    center : boolean
+        If True, pads beginning of signal so timestamps correspond
+        to center of window.
+
+    Returns
+    -------
+    input_data (np.ndarray): The preprocessed audio. Depending on 
+        the value of input_repr, it will be np.ndarray[batch, time, frequency, 1]
+        if a valid input representation is provided,
+        or np.ndarray[batch, time, 1] if no input_repr is provided.
+    """
+    # if audio is shaped like a spectrogram:  # if shaped like [freq, time], convert to [batch, freq, time, channel]
+    #     stft_hop_size = 242  # needs to know stft hop size
+    #     return librosa.util.frame(x, int(sr/stft_hop_size), int(sr/stft_hop_size*hop_size)).transpose((2, 0, 1))[...,None]
+    x = _preprocess_audio_batch(audio, sr, hop_size=hop_size, center=center)
+    if input_repr:
+        if input_repr == 'linear':
+            x = np.stack([_librosa_linear_frontend(xi[0], **kw) for xi in x])[...,None]
+        elif input_repr == 'mel128':
+            x = np.stack([_librosa_mel_frontend(xi[0], n_mels=128, **kw) for xi in x])[...,None]
+        elif input_repr == 'mel256':
+            x = np.stack([_librosa_mel_frontend(xi[0], n_mels=256, **kw) for xi in x])[...,None]
+        else:
+            raise OpenL3Error('Invalid input representation "{}"'.format(input_repr))
+    return x
+
+
+
+def get_audio_embedding(audio, sr=48000, model=None, input_repr="mel256",
                         content_type="music", embedding_size=6144,
                         center=True, hop_size=0.1, batch_size=32,
-                        verbose=True):
+                        frontend='auto', verbose=True):
```

Comment:
You've asked about this is numerous places so I guess I'll just pick one spot to respond haha.

To answer your question above (just in case), if the frontend is set to auto, it will look at the model input shape. If it has 3 dimensions i.e. `(batch, channel, time)` then it takes PCM, but if it has 4 dimensions i.e. `(batch frequency, time, channel)` then it takes spectrograms. If there wasn't a clear distinction between the two input types then I would be much less eager to try to infer it.

So I'm going to push back a bit here. 

I really don't see a scenario where detecting the frontend would introduce bugs and one could argue that having to specify the same thing in two places could lead to more bugs/frustration because it's one more variable that they have to think about and keep track of.

Personally I don't like making APIs unnecessarily verbose and I don't think it would protect against someone wanting to use kapre and accidentally using librosa, for example.

For instance:
```python
def process_things(files, ..., frontend='librosa'):
    model = openl3.load_audio_embedding_model(..., frontend=frontend)
    for f in files:
        openl3.process_audio_file(f, model=model, frontend=frontend)
        # vs 
        openl3.process_audio_file(f, model=model)

```
I don't see how the first option is safer than the second. And if `process_audio_file` gets refactored to somewhere else, then they not only have to track the model, but they have to keep an explicit record of their frontend along side the model. Whereas if we can detect it, then if they want to switch from kapre to librosa, they only have to make the change in one place.

I do admit that I'm not in love with adding the additional `'auto'` value, but I think it's the most clear way to express it. An alternative could be `'infer'` which is a bit more clear about what's happening. 
(sidenote: usually when I want something to be calculated internally, I use `None` as a `null`/do-the-default-thing value, but I think that its falsey-ness would confuse things here.)

To propose a solution:
I will add some validation checks in `_validate_audio_frontend` so that if a user passes an explicit frontend that doesn't match the expected shape of the model, it will throw an error (basically if they pass `frontend=kapre` with a model that doesn't have a 3d input, then we throw an error, and same with librosa/4d).

So if someone wants to ensure that they are using the correct frontend it will throw an error if it's different, but otherwise we should just assume that the user knows what they are doing and assume that when they pass a librosa-compatible model, that they want to use a librosa frontend.

Reply:
Actually one thing that I noticed - when using the librosa frontend, `input_repr` is confusing because on one hand, it is a required argument for `load_audio_embedding_model` (and the first argument as well) but it doesn't really do anything.

In order for librosa to consider your specified `input_repr`, you have to pass it to `get_audio_embedding` as well.

So to fix that, I'm going to throw an error if they don't specify `input_repr` with the librosa frontend. That's something that we can't detect from the model.

I just wish it wasn't a required argument to load the model because it's confusing.